treescript/src/treescript/github/client.py

```diff
@@ -0,0 +1,138 @@
+"""Treescript git functions."""
+import base64
+import logging
+from collections import defaultdict
+import os
+from string import Template
+from textwrap import dedent
+from typing import Dict, List, Optional, Union
+
+from simple_github import AppClient
+
+log = logging.getLogger(__name__)
+
+GITHUB_APP_ID = 390757  # https://github.com/apps/releng-treescript
+
+
+class GithubClient:
+    def __init__(self, config, owner, repo):
+        with open(config["github_config"]["privkey_file"]) as fh:
+            privkey = fh.read()
+        self.owner = owner
+        self.repo = repo
+        self._client = AppClient(GITHUB_APP_ID, privkey, owner=owner, repositories=[repo])
+
+    async def close(self):
+        await self._client.close()
+
+    async def __aenter__(self):
+        return self
+
+    async def __aexit__(self, *excinfo):
+        await self.close()
+
+    async def commit(self, branch: str, message: str, additions: Optional[Dict[str, str]] = None, deletions: Optional[List[str]] = None) -> None:
+        """Commit changes to the given repository and branch.
+
+        Args:
+            branch (str): The branch name to commit to.
+            message (str): The commit message to use.
+            additions (Dict): Files to add or update in the commit. Of the form
+                `{<path>: <contents>}` (optional).
+            deletions (List): Files to delete in the commit. Of the form
+                `[<path>]` (optional).
+        """
+        changes = defaultdict(list)
+        if additions:
+            for name, contents in additions.items():
+                changes["additions"].append({"path": name, "contents": base64.b64encode(contents.encode("utf-8")).decode("utf-8")})
+
+        if deletions:
+            changes["deletions"] = [{"path": p} for p in deletions]
+
+        if not changes:
+            log.warn("No changes to commit, aborting.")
+            return
+
+        query = Template(
```

Comment:
Question: TIL string templates! What are the pros of them compared to an f-string or a `str.format()` call?

Reply:
Oops, missed this comment.

The main reason I'm using them is because `f-string` and `str.format` treat `{}` as special characters. Since GraphQL is JSON-like and uses a lot of curly braces, it would need a lot of escaping. This just makes the template look a little nicer imo.